internal/worker/firewaller/firewaller.go

```diff
@@ -580,30 +585,35 @@ func (fw *Firewaller) startUnit(ctx context.Context, unit Unit, machineTag names
 // startApplication creates a new data value for tracking details of the
 // application and starts watching the application for exposure changes.
 func (fw *Firewaller) startApplication(ctx context.Context, app Application) error {
-	exposed, exposedEndpoints, err := app.ExposeInfo(ctx)
+	isExposed, err := fw.applicationService.IsApplicationExposed(ctx, app.Name())
 	if err != nil {
-		return err
+		return internalerrors.Capture(err)
+	}
+	exposedEndpoints, err := fw.applicationService.GetExposedEndpoints(ctx, app.Name())
+	if err != nil {
+		return internalerrors.Capture(err)
 	}
 	applicationd := &applicationData{
-		fw:               fw,
-		application:      app,
-		exposed:          exposed,
-		exposedEndpoints: exposedEndpoints,
-		unitds:           make(map[coreunit.Name]*unitData),
+		fw:                 fw,
+		applicationTag:     app.Tag(),
+		applicationService: fw.applicationService,
+		exposed:            isExposed,
+		exposedEndpoints:   exposedEndpoints,
+		unitds:             make(map[coreunit.Name]*unitData),
 	}
 	fw.applicationids[app.Tag()] = applicationd
```

Comment:
@manadart this should be using a runner. We should fix this at some point. The `applicationids` map can get out of sync and **might** (I'd have to check, it looks like it, but the code isn't exactly simple) not always be accessed in the same goroutine.